File: src/main/java/dev/rolypolyvole/golemtargets/util/GolemTargetGoal.java
package dev.rolypolyvole.golemtargets.util;

import dev.rolypolyvole.golemtargets.config.GolemTargetsConfig;
import net.minecraft.core.component.DataComponents;
import net.minecraft.world.SimpleContainer;
import net.minecraft.world.entity.ai.goal.target.TargetGoal;
import net.minecraft.world.entity.animal.golem.AbstractGolem;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;

import java.util.HashSet;
import java.util.Set;

public class GolemTargetGoal extends TargetGoal {

    private final AbstractGolem golem;

    public GolemTargetGoal(AbstractGolem golem) {
        super(golem, false);
        this.golem = golem;
    }

    @Override
    public boolean canUse() {
        SimpleContainer container = ((GolemTargetAccessor) golem).golemTargets$getContainer();
        Set<String> targetNames = new HashSet<>();

        for (int i = 0; i < container.getContainerSize(); i++) {
            ItemStack stack = container.getItem(i);
            if (!stack.isEmpty() && stack.is(Items.PAPER) && stack.has(DataComponents.CUSTOM_NAME)) {
                targetNames.add(stack.get(DataComponents.CUSTOM_NAME).getString());
            }
        }

        if (targetNames.isEmpty()) return false;

        double rangeSq = GolemTargetsConfig.range * GolemTargetsConfig.range;

        Player closest = null;
        double closestDist = Double.MAX_VALUE;
        for (Player player : golem.level().players()) {
            if (player.isSpectator() || player.isCreative()) continue;
            if (!targetNames.contains(player.getName().getString())) continue;
            double dist = player.distanceToSqr(golem);
            if (dist > rangeSq) continue;
            if (dist < closestDist) {
                closestDist = dist;
                closest = player;
            }
        }

        this.targetMob = closest;
        return this.targetMob != null;
    }

    @Override
    public boolean canContinueToUse() {
        if (this.targetMob == null || !this.targetMob.isAlive()) return false;
        if (this.targetMob instanceof Player player && (player.isSpectator() || player.isCreative())) return false;

        double rangeSq = GolemTargetsConfig.range * GolemTargetsConfig.range;
        if (this.targetMob.distanceToSqr(golem) > rangeSq) return false;

        SimpleContainer container = ((GolemTargetAccessor) golem).golemTargets$getContainer();
        String targetName = this.targetMob.getName().getString();
        for (int i = 0; i < container.getContainerSize(); i++) {
            ItemStack stack = container.getItem(i);
            if (!stack.isEmpty() && stack.is(Items.PAPER) && stack.has(DataComponents.CUSTOM_NAME)) {
                if (stack.get(DataComponents.CUSTOM_NAME).getString().equals(targetName)) {
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    public void start() {
        this.mob.setTarget(this.targetMob);
    }

    @Override
    public void tick() {
        this.mob.setTarget(this.targetMob);
    }

    @Override
    public void stop() {
        this.targetMob = null;
        this.mob.setTarget(null);
    }
}

File: src/main/kotlin/dev/rolypolyvole/golemtargets/GolemTargetGoal.kt
package dev.rolypolyvole.golemtargets

import net.minecraft.core.component.DataComponents
import net.minecraft.world.entity.ai.goal.target.TargetGoal
import net.minecraft.world.entity.animal.golem.AbstractGolem
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Items

class GolemTargetGoal(private val golem: AbstractGolem) : TargetGoal(golem, false) {

    override fun canUse(): Boolean {
        val container = (golem as GolemTargetAccessor).`golemTargets$getContainer`()
        val targetNames = mutableSetOf<String>()

        for (i in 0 until container.containerSize) {
            val stack = container.getItem(i)
            if (!stack.isEmpty && stack.`is`(Items.PAPER) && stack.has(DataComponents.CUSTOM_NAME)) {
                targetNames.add(stack.get(DataComponents.CUSTOM_NAME)!!.string)
            }
        }

        if (targetNames.isEmpty()) return false

        targetMob = golem.level().players()
            .filter { it.name.string in targetNames && !it.isSpectator && !it.isCreative }
            .minByOrNull { it.distanceToSqr(golem) }

        return targetMob != null
    }

    override fun canContinueToUse(): Boolean {
        val target = targetMob ?: return false
        if (!target.isAlive) return false
        if (target is Player && (target.isSpectator || target.isCreative)) return false

        val container = (golem as GolemTargetAccessor).`golemTargets$getContainer`()
        val targetName = target.name.string
        for (i in 0 until container.containerSize) {
            val stack = container.getItem(i)
            if (!stack.isEmpty && stack.`is`(Items.PAPER) && stack.has(DataComponents.CUSTOM_NAME)) {
                if (stack.get(DataComponents.CUSTOM_NAME)!!.string == targetName) {
                    return true
                }
            }
        }
        return false
    }

    override fun start() {
        mob.target = targetMob
    }

    override fun tick() {
        mob.target = targetMob
    }

    override fun stop() {
        targetMob = null
        mob.target = null
    }
}
